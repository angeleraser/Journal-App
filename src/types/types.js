export const TYPES = {
  login: "[Auth] Login",
  logout: "[Auth] Logout",

  uiSetError: "[UI] Set Error",
  uiRemoveError: "[UI] Remove Error",

  uiStartLoading: "[UI] Start Loading",
  uiFinishLoading: "[UI] Finish Loading",

  notesAddNew: "[Notes] Add new note",
  notesSelect: "[Notes] Select note",
  notesLoad: "[Notes] Load notes",
  notesUpdate: "[Notes] Update note",
  notesFileURL: "[Notes] Update image url",
  notesDelete: "[Notes] Delete note",
  notesLogoutCleaning: "[Notes] Logout cleaning",
};
