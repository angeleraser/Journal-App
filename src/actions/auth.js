import Swal from "sweetalert2";
import { FIREBASE, googleAuthProvider } from "..";
import { TYPES } from "../types/types";
import { uiStartLoading, uiFinishLoading } from "./ui";
// NOTA: Si la tarea es asincrona la funcion debe retonar otra funcion que realiza el dispatch, sino un objeto con el tipo de accion y el payload normal.

export const login = (uID, displayName) => ({
  type: TYPES.login,
  payload: {
    uID,
    displayName,
  },
});

export const startLoginEmailPassword = (email, password) => {
  return async (dispatch) => {
    try {
      dispatch(uiStartLoading());
      {
        const { user } = await FIREBASE.auth().signInWithEmailAndPassword(
          email,
          password
        );
        dispatch(login(user.uid, user.displayName));
      }
    } catch (error) {
      Swal.fire("Error", error.message, "error");
      console.log(error);
    } finally {
      dispatch(uiFinishLoading());
    }
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return async (dispatch) => {
    try {
      dispatch(uiStartLoading());
      {
        const { user } = await FIREBASE.auth().createUserWithEmailAndPassword(
          email,
          password
        );
        await user.updateProfile({ displayName: name });
        dispatch(login(user.uid, user.displayName));
      }
    } catch (error) {
      console.log(error);
      Swal.fire("Error", error.message, "error");
    } finally {
      dispatch(uiFinishLoading());
    }
  };
};

export const startGoogleLogin = () => {
  return async (dispatch) => {
    try {
      dispatch(uiStartLoading());
      {
        const { user } = await FIREBASE.auth().signInWithPopup(
          googleAuthProvider
        );
        dispatch(login(user.id, user.displayName));
      }
    } catch (error) {
      console.log(error);
      Swal.fire("Error", error.message, "error");
    } finally {
      dispatch(uiFinishLoading());
    }
  };
};

export const startLogout = () => {
  return async (dispatch) => {
    await FIREBASE.auth().signOut();
    dispatch(logout());
  };
};

export const logout = () => ({
  type: TYPES.logout,
});
