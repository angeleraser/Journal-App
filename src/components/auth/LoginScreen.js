import React from "react";
import { Link } from "react-router-dom";
import { googleIcon } from "../../icons/svg";
import useForm from "../hooks/useForm";
import { useDispatch, useSelector } from "react-redux";
import { startGoogleLogin, startLoginEmailPassword } from "../../actions/auth";
export const LoginScreen = () => {
  const [{ email, password }, handleInputChange] = useForm({
    email: "",
    password: "",
  });
  const dispatch = useDispatch();
  const { loading } = useSelector(({ ui }) => ui);
  const handleLogin = (e) => {
    e.preventDefault();
    dispatch(startLoginEmailPassword(email, password));
  };
  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  };
  return (
    <>
      <h3 className="auth__title">Login</h3>
      <form
        onSubmit={handleLogin}
        autoComplete="off"
        className="flex flex-col w-full items-start">
        <input
          className="auth__input"
          type="email"
          name="email"
          placeholder="email@example.com"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          name="password"
          placeholder="Enter your password..."
          className="auth__input"
          value={password}
          onChange={handleInputChange}
        />
        <button
          disabled={loading}
          className="btn btn-primary btn-block"
          type="submit">
          {!loading ? "Login" : "Loading..."}
        </button>
        <div className="auth__social-networks">
          <div className="networks-buttons w-full mb-4">
            <button
              type="button"
              onClick={handleGoogleLogin}
              className="google-btn-wrapper w-full flex items-center">
              <div className="google-icon mr-10"> {googleIcon}</div>
              Sign in with Google
            </button>
          </div>
        </div>
        <Link className="link font-bold w-full text-center" to="/auth/register">
          Create new account
        </Link>
      </form>
    </>
  );
};
