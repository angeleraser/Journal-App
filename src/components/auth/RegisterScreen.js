import React from "react";
import { Link } from "react-router-dom";
import useForm from "../hooks/useForm";
import { useDispatch, useSelector } from "react-redux";
import { isFormValid } from "../../helpers/isFormValid.js";
import { startRegisterWithEmailPasswordName } from "../../actions/auth";
export const RegisterScreen = () => {
  const [
    { name, email, password, confirmPassword },
    handleInputChange,
  ] = useForm({
    name: "",
    email: "",
    password: "",
    confirmPassword: "",
  });
  const dispatch = useDispatch();
  const { errorMsg } = useSelector((state) => state.ui);
  const handleRegister = (e) => {
    e.preventDefault();
    if (
      isFormValid(
        { name, email, password, confirmPassword },
        dispatch,
        errorMsg
      )
    ) {
      dispatch(startRegisterWithEmailPasswordName(email, password, name));
    }
  };
  return (
    <>
      {" "}
      <h3 className="auth__title">Register</h3>
      {!!errorMsg && (
        <p className="auth__invalid-form-msg w-full text-center font-bold mb-4">
          {" "}
          {errorMsg}
        </p>
      )}
      <form
        onSubmit={handleRegister}
        autoComplete="off"
        className="flex flex-col w-full items-start">
        <input
          className="auth__input"
          type="name"
          value={name}
          name="name"
          onChange={handleInputChange}
          placeholder="Your name"
        />
        <input
          className="auth__input"
          value={email}
          type="email"
          name="email"
          onChange={handleInputChange}
          placeholder="email@example.com"
        />
        <input
          type="password"
          name="password"
          value={password}
          onChange={handleInputChange}
          placeholder="Enter your password..."
          className="auth__input"
        />
        <input
          type="password"
          value={confirmPassword}
          name="confirmPassword"
          onChange={handleInputChange}
          placeholder="Confirm your password..."
          className="auth__input"
        />
        <button className="btn mb-6 btn-primary btn-block" type="submit">
          Register
        </button>
        <Link className="link" to="/auth/login">
          Already register?
        </Link>
      </form>
    </>
  );
};
