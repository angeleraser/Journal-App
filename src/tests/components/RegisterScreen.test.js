import React from "react";
import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom";
import { RegisterScreen } from "../../components/auth/RegisterScreen";
import { TYPES } from "../../types/types";
// Mock
// jest.mock("../../actions/auth", () => {
//   return {
//     startGoogleLogin: jest.fn(),
//     startLoginEmailPassword: jest.fn(),
//   };
// });

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
// Hay que definir el estado inicial igual al de la aplicacion en ese punto
const initState = {
  auth: {},
  notes: {
    notes: [],
    selected: null,
  },
  ui: {
    loading: false,
    errorMsg: null,
  },
};
let store = mockStore(initState);
// Fix dispatch del store si las acciones a evaluar son asincronas
// store.dispatch = jest.fn();
const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <RegisterScreen />
    </MemoryRouter>
  </Provider>
);
describe("Testing Login Screen", () => {
  test("should match to snap", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should dispatch setError action", () => {
    wrapper.find("form").simulate("submit", {
      preventDefault() {},
    });
    const actions = store.getActions();
    expect(actions.some((action) => action.type === TYPES.uiSetError));
  });

  test("should render msgError HTML element", () => {
    const initState = {
      ui: {
        loading: false,
        errorMsg: "Please enter a valid email",
      },
    };
    let store = mockStore(initState);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <RegisterScreen />
        </MemoryRouter>
      </Provider>
    );
    const msgErrorElement = wrapper.find(".auth__invalid-form-msg");
    expect(msgErrorElement.exists()).toBe(true);
    expect(msgErrorElement.text().trim()).toBe("Please enter a valid email");
  });
});
