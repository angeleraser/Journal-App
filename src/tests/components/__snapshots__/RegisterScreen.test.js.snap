// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Login Screen should match to snap 1`] = `
Array [
  " ",
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    autoComplete="off"
    className="flex flex-col w-full items-start"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Your name"
      type="name"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="email@example.com"
      type="email"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Enter your password..."
      type="password"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="confirmPassword"
      onChange={[Function]}
      placeholder="Confirm your password..."
      type="password"
      value=""
    />
    <button
      className="btn mb-6 btn-primary btn-block"
      disabled={false}
      type="submit"
    >
      Register
    </button>
    <a
      className="link font-bold w-full text-center"
      href="/auth/login"
      onClick={[Function]}
    >
      Already register?
    </a>
  </form>,
]
`;
